### PR TITLE
Field conversions

### PR DESCRIPTION
**Pull request by mjrodgers**
Add parent object overload conversion methods for vectors of integers (or ZZRingElems) into finite field elements.
This works for `fpField`, `FpField`, `fqPolyRepField`, `FqPolyRepField`.
Currently these methods are supplied in OSCAR only for `fpField`/`fqPolyRepField`s supplied by `experimental/GModule`. But it seems like these belong in Nemo.

### CODE REVIEW

**Comment by mjrodgers** [comment]
I'm running into some issues because a few of the functions I'm relying on are implemented in Hecke and don't seem to be exposed in Nemo. Should I be submitting this pull request to Hecke instead of Nemo?

**Comment by mjrodgers** [comment]
I think the issue is only at the moment with `fqPolyRepFieldElem`; I can get around the issue with `FqPolyRepFieldElem` since I can call `FqPolyRepFieldElem(ctx::FqPolyRepField, b::ZZPolyRingElem)`, but the equivalent function does not exist for small `fq` (it has in its place `fqPolyRepFieldElem(ctx::fqPolyRepField, b::Vector{UInt64})`).

**Comment by thofma** [comment]
Which functions do you need?

**Comment by thofma** [comment]
I suggest to add two litte helper functions
```
function _some_good_name(R::FqPolyRepField, x::FpPolyRingElem)
  z = R()
  ccall((:fq_set_fmpz_mod_poly, libflint), Nothing, (Ref{Nemo.FqPolyRepFieldElem}, Ref{Nemo.FpPolyRingElem}, Ref{FqPolyRepField}), z, x, R)
  return z
end

function _some_good_name(A::fqPolyRepField, x::fpPolyRingElem)
  u = A()
  ccall((:fq_nmod_set, libflint), Nothing,
                     (Ref{fqPolyRepFieldElem}, Ref{fpPolyRingElem}, Ref{fqPolyRepField})
                                     u, x, A)
  return u
end
```
These are slightly different from the ones in Hecke, since some flint functions have been fixed since then.

**Comment by mjrodgers** [comment]
Sorry if this is the wrong approach, but trying figure out through trial and error how to contribute in the way that makes the most sense.

Does it make sense at all to put these helper functions in with the FlintTypes? So for example, inside the struct for `fqPolyRepFieldElem`, adding
```
   function fqPolyRepFieldElem(ctx::fqPolyRepField, x::fpPolyRingElem)
      d = new()
      ccall((:fq_nmod_init2, libflint), Nothing,
            (Ref{fqPolyRepFieldElem}, Ref{fqPolyRepField}), d, ctx)
      finalizer(_fq_nmod_clear_fn, d)
      ccall((:fq_nmod_set, libflint), Nothing,
            (Ref{fqPolyRepFieldElem}, Ref{fpPolyRingElem}, Ref{fqPolyRepField})
                  d, x, ctx)
      return d
   end
   ```

**Comment by thofma** [comment]
Ah, yes, perfect, this makes the most sense.

**Comment by thofma** [comment]
You have to be a bit careful. If you `fq_nmod_set` directly, it must be reduced with respect to the defining polynomial. Otherwise use `fq_nmod_set_nmod_poly`.

**Comment by mjrodgers** [comment]
Okay, that makes sense; I will use `_poly`.

Is there a reason that the `FqPolyRepFieldElem` versions functions have a final step (compared to `fq` versions) setting `d.parent = ctx`? I will match for consistency, but just curious why this is needed here.

**Comment by thofma** [comment]
In the C library, the elements do not know their parent. (All of the functions thus require the ctx object).

In Nemo, in addition we store the parent/ctx directly on the element. If the object is constructed via `new()` in a constructor, then we have to set the `ctx` separately. If an object is created via `R()`, where `R` is the parent, then the `ctx` is already set correctly.

I hope I understood your question correctly.

**Comment by mjrodgers** [comment]
This makes sense, but the question is then why the small `fq` versions are missing this line to set the parent, while all the big `Fq` versions include it.

**Comment by thofma** [comment]
Where is the discrepancy? Can you link to the line here?

**Comment by mjrodgers** [comment]
The `fqPolyRepFieldElem` functions start on [line 2300](https://github.com/Nemocas/Nemo.jl/blob/master/src/flint/FlintTypes.jl#L2300) and do not set the parent.

The `FqPolyRepFieldElem` functions start on [line 2764](https://github.com/Nemocas/Nemo.jl/blob/master/src/flint/FlintTypes.jl#L2300) and all set the parent.

**Comment by mjrodgers** [comment]
Interesting, it looks like most of the functions that use these internal methods to create field elements set the parent themselves even though it is (often) redundant, see [here for example](https://github.com/Nemocas/Nemo.jl/blob/master/src/flint/fq.jl#L594). But if this isn't done for the `fq` fields (like in my latest commit, oops),  you get a problem with (I'm sure among other things) comparing `==` because there is no parent to compare.

Since it seems like all of the functions that use these methods DO in fact set the parent themselves, this isn't really an issue except for consistency between `fq`/`Fq` fields. Unless there is some performance hit from setting this when it doesn't need to be, I doubt that though. And with the new FqDefault it will maybe be a relic of the past anyway.

File: src/flint/fq.jl [hunk 1]
```diff
@@ -620,6 +620,14 @@ function (a::FqPolyRepField)(b::FqPolyRepFieldElem)
     end
 end
 
+function (a::FqPolyRepField)(b::Vector{<:IntegerUnion})
+   da = degree(a)
+   db = length(b)
+   da < db && error("Coercion impossible")
+   F = GF(characteristic(a))
```
Comment (thofma) [inline src/flint/fq.jl:627]:
```suggestion
   F = GF(characteristic(a), cached = false)
```

File: src/flint/fq_nmod.jl [hunk 2]
```diff
@@ -665,6 +665,16 @@ function (a::fqPolyRepField)(b::fqPolyRepFieldElem)
     end
 end
 
+function (a::fqPolyRepField)(b::Vector{<:IntegerUnion})
+   da = degree(a)
+   db = length(b)
+   da < db && error("Coercion impossible")
+   F = GF(Int(characteristic(a)))
```
Comment (thofma) [inline src/flint/fq_nmod.jl:672]:
```suggestion
   F = GF(Int(characteristic(a)), cached = false)
```

File: src/flint/gfp_elem.jl [hunk 3]
```diff
@@ -469,6 +469,11 @@ function (R::fpField)(a::Union{fpFieldElem, zzModRingElem, FpFieldElem, ZZModRin
    end
 end
 
+function (R::fpField)(a::Vector{<:IntegerUnion})
+   isone(length(a)) || error("Coercion impossible")
```
Comment (thofma) [inline src/flint/gfp_elem.jl:473]: Any particular reason why you do not  allow length zero? You allow it in the `fqPolyRepField` case.

**Comment by thofma** [comment]
@fieker Do you have an opinion on allow
```
julia> F = GF(2, 3);

julia> F([1, 1])
```
that is, allowing the number of coordinates being less than the absolute degree.

File: src/flint/gfp_elem.jl (hunk 3 above)
Comment (mjrodgers) [reply src/flint/gfp_elem.jl:473]: Ah this was a simple oversight.  Fixed for now (unless Claus wants to require `length(a) == degree(R)`)

**Comment by mjrodgers** [comment]
This last failing required test looks to be unrelated to my pull request (some doctests in `Oscar/src/Rings/` and `Oscar/src/Modules/`).

**Pull request merged**